src/CowWrapper.sol

```diff
@@ -0,0 +1,207 @@
+///      - Implement _wrap() for custom logic
+abstract contract CowWrapper is ICowWrapper {
+    /// @notice Thrown when the caller is not an authenticated solver
+    /// @param unauthorized The address that attempted to call wrappedSettle
+    error NotASolver(address unauthorized);
+
+    /// @notice Thrown when settle data doesn't contain the correct function selector
+    /// @param invalidSettleData The invalid settle data that was provided
+    error InvalidSettleData(bytes invalidSettleData);
+
+    /// @notice The settlement contract
+    ICowSettlement public immutable SETTLEMENT;
+
+    /// @notice The authentication contract used to verify solvers
+    /// @dev This is derived from `SETTLEMENT.authenticator()`.
+    ICowAuthentication public immutable AUTHENTICATOR;
+
+    /// @notice Constructs a new CowWrapper
+    /// @param settlement_ The ICowSettlement contract to use at the end of the wrapper chain. Also used for wrapper authentication.
+    constructor(ICowSettlement settlement_) {
+        SETTLEMENT = settlement_;
+        AUTHENTICATOR = settlement_.authenticator();
+    }
+
+    /// @notice Initiates a wrapped settlement call
+    /// @dev Entry point for solvers to execute wrapped settlements. Verifies the caller is a solver,
+    ///      validates wrapper data, then delegates to _wrap() for custom logic.
+    /// @param settleData ABI-encoded call to ICowSettlement.settle() containing trade data
+    /// @param wrapperData Encoded data for this wrapper and the chain of next wrappers/settlement.
+    ///                    Format: [2-byte len][wrapper-specific-data][next-address]([2-byte len][wrapper-specific-data][next-address]...)
+    function wrappedSettle(bytes calldata settleData, bytes calldata wrapperData) external {
+        // Revert if not a valid solver
+        require(AUTHENTICATOR.isSolver(msg.sender), NotASolver(msg.sender));
+
+        // Find out how long the next wrapper data is supposed to be
+        // We use 2 bytes to decode the length of the wrapper data because it allows for up to 64KB of data for each wrapper.
+        // This should be plenty of length for all identified use-cases of wrappers in the forseeable future.
+        uint16 nextWrapperDataLen = uint16(bytes2(wrapperData[0:2]));
+
+        // Delegate to the wrapper's custom logic
+        uint256 remainingWrapperDataStart = 2 + nextWrapperDataLen;
+        _wrap(settleData, wrapperData[2:remainingWrapperDataStart], wrapperData[remainingWrapperDataStart:]);
+    }
+
+    /// @notice Parses and validates wrapper-specific data
+    /// @dev Must be implemented by concrete wrapper contracts. Used for pre-execution validation.
+    ///      The implementation should consume its wrapper-specific data and return the remainder.
+    /// @param wrapperData The full wrapper data to parse
+    /// @return remainingWrapperData The portion of wrapper data not consumed by this wrapper
```

Comment:
In this case (and a few others) you can avoid duplication by simply inheriting them from the interface:

```suggestion
    /// @inheritdoc ICowWrapper
```